src/main/java/com/iota/iri/controllers/TransactionViewModel.java

```diff
@@ -379,98 +379,76 @@ public static void updateSolidTransactions(Tangle tangle, final Set<Hash> analyz
+     * This method calculates the correct referenced milestone and updates the transaction.
+     * The referenced milestone is recursively defined by {@link #chooseReferencedMilestone}
      *
      * @param tangle database interface
      * @throws Exception if the database update of the database fails
+     * @return the referenced milestone
      */
-    public void updateReferencedSnapshot(Tangle tangle) throws Exception {
-        // trigger the retrieval which will also cause an update
-        this.referencedSnapshot(tangle);
+    private int findReferencedMilestone(Tangle tangle) throws Exception {
+        // if the referenced snapshot is 0 -> we try to recursively determine it
+        if(transaction.referencedMilestone == 0) {
+            TransactionViewModel trunk = this.getTrunkTransaction(tangle);
+            TransactionViewModel branch = this.getBranchTransaction(tangle);
+
+            int trunkReferencedSnapshot = findReferencedMilestoneRecursively(trunk, tangle);
+            int branchReferencedSnapshot = findReferencedMilestoneRecursively(branch, tangle);
+
+            return chooseReferencedMilestone(trunkReferencedSnapshot, branchReferencedSnapshot);
+        }
+
+        // else return the stored value
+        return transaction.referencedMilestone;
     }
 
     /**
-     * This method allows to set the referenced snapshot for this transaction.
      *
-     * It first checks, if the value is already set and issues a database update if the value differs from the current
-     * one.
-     *
-     * @param tangle database interface
-     * @param referencedSnapshot milestone index that is referenced by this transaction
-     * @throws Exception if the database update of the database fails
+     * @param trunkReferencedMilestone - the referenced milestone of the trunk
+     * @param branchReferencedMilestone - the referenced milestone of the branch
+     * @return the max value of the two params
      */
-    public void referencedSnapshot(Tangle tangle, final int referencedSnapshot) throws Exception {
-        if ( referencedSnapshot != transaction.referencedSnapshot ) {
-            transaction.referencedSnapshot = referencedSnapshot;
-            update(tangle, "referencedSnapshot");
+    public static int chooseReferencedMilestone(int trunkReferencedMilestone, int branchReferencedMilestone) {
+        return Math.max(trunkReferencedMilestone, branchReferencedMilestone);
+    }
+
+    private static int findReferencedMilestoneRecursively(TransactionViewModel approvee, Tangle tangle) throws Exception {
+        int referencedMilestone = approvee.getReferencedMilestone();
+        if(referencedMilestone == 0 && !Objects.equals(approvee.getHash(), Hash.NULL_HASH)) {
+            // check if the referenced transaction is a milestone -> otherwise descend recursively
+            try {
+                MilestoneViewModel milestoneViewModel = MilestoneViewModel.fromHash(tangle, approvee.getHash());
```

Comment:
It is better to just check whether it is coo address.
We need the configuration object for this...

Also there is a problem that we replicate `TransactionViewModel` a lot so it is not good to have it take more space.
In the future design I have in my mind we won't be replicating it like we do today.